File: kmeans.c
# include <stdio.h>
#include <assert.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
int mainFuncV2 (int K,  int max_iter);
double CreateVector();
double distanceFromMean(int i1, int j1, int n,double vec [], double mean [] );
void updateMean (int j, int place, int n,double vec [], double mean [], int numOfVec);
void updateCheck (int n,double vec [], double mean []);
int toStop (int n,double vec [], double mean []);

int main (int argc, char *argv[])
{
    if(argc==2){
            mainFuncV2 (atoi(argv[1]),200); }
    if(argc==3){
            mainFuncV2 (atoi(argv[1]),atoi(argv[2])); }
    assert(argc==2||argc==3);
    return (0);
}


double distanceFromMean(int i1, int j1, int n,double vec [], double mean [] ){
    double d=0;
    double meanI;
    double x;
    double distance;
    int i;
    for ( i = 0; i < n; i++)
    {
        meanI= mean[n*j1+i];
        x=vec[n*i1+i];
        distance= (x-meanI)*(x-meanI);
        d=d+distance;         
    }
    return (d);
}

void updateMean (int j, int place, int n,double vec [], double mean [], int numOfVec){
    int i;
    for ( i = 0; i < n; i++)
    {
        double meanI= mean[i+place*n];
        double sum;
        if (numOfVec>0){
            sum= meanI*numOfVec+vec[i+j*n];
        }
        else{
            sum= vec[i+j*n];
        }
        mean[i+place*n]= sum/(numOfVec+1); 
    }
}
void updateCheck (int n,double vec [], double mean []){
    int i;
    for ( i = 0; i < n; i++)
    {
        vec[i]=mean[i];
    }   
}

int toStop (int n,double vec [], double mean []){
    int i;
    for ( i = 0; i < n; i++){
        if(vec[i]!=mean[i]){
            return (0);
        }
    }
    return (1);
}


int mainFuncV2 (int K,  int max_iter) {
int i,j,cc,row,columns,b1,k1,k2,lines,iter,place,d,dim,i1,j1;
double distance,curDistance,cur;
char * token;
double *vecOfMeansOld;
double *vecOfMeansNew;
double *vector;
char *singleLine;
int *numberOfVecPerGroup;
char line[1000];
assert(K>0);
assert(max_iter>0);
numberOfVecPerGroup=(int *)malloc(sizeof(int)*K);
assert(numberOfVecPerGroup!=NULL);
lines=0;
i=0;
dim=1;
    while (scanf("%s",line)!=EOF){
        if (dim==1){
            while (line[i]!='\0'){
                if (line[i]==','){
    
                    dim++;
                }
                i++;
            }
        }
            lines++;
        }
    assert(K<lines);
    rewind(stdin);
    vector = (double *)malloc(sizeof(double)*(lines+1)*(dim+1));
    assert(vector!=NULL);
    i1=0;
    j1=0;
    singleLine=(char *)malloc(sizeof(char)*dim*64);
    assert(singleLine!=NULL);
    while(scanf("%s",singleLine)!=EOF){
        assert(singleLine!=NULL);
        token = strtok(singleLine, ",");
        while( token != NULL ) {
            cur=atof(token);
            vector[dim*i1+j1]=cur;
            token = strtok(NULL, ",");
            j1++;
        }
        i1++;
        j1=0;
    }
free(singleLine);
iter=0;
for ( iter = 0; iter < K; iter++)
{
    numberOfVecPerGroup[iter]=0;

}
iter=0;
vecOfMeansOld = (double *)malloc(sizeof(double)*(K+1)*(dim+1));
assert(vecOfMeansOld!=NULL);
vecOfMeansNew = (double *)malloc(sizeof(double)*(K+1)*(dim+1));
assert(vecOfMeansNew!=NULL);

for ( i = 0; i < K; i++)
{
    for ( cc = 0; cc < dim; cc++)
    {
        vecOfMeansOld[i*dim+cc]= vector[i*dim+cc];
        
    }
    
}
d=dim;
while(iter<max_iter){
        
        for ( i = 0; i < lines ; i++)
        {
            distance=0;
            place=0;
            
            for (  j = 0; j < K; j++)
            {
                curDistance= distanceFromMean(i,j, d, vector, vecOfMeansOld );
                if (j==0){
                        distance=curDistance;
                }
                if(curDistance< distance){
                    distance=curDistance;
                    place=j;
                } 
            }
            updateMean (i,place,d, vector, vecOfMeansNew , numberOfVecPerGroup[place]);
            numberOfVecPerGroup[place]=numberOfVecPerGroup[place]+1;
        }

        b1=1;
        for (k1=0; k1<K; k1++)
        {
            for (k2 = 0; k2 < dim; k2++)
            {
                if(vecOfMeansOld[k1*dim+k2]!=vecOfMeansNew[k1*dim+k2]){
                    b1=0;
                }
            }    
        }
        if (b1==1){
            break;
        }
        for (k1=0; k1<K; k1++)
        {
            for (k2 = 0; k2 < dim; k2++)
            {
                vecOfMeansOld[k1*dim+k2]=vecOfMeansNew[k1*dim+k2];
                vecOfMeansNew[k1*dim+k2]=0;
            }
            
        }
        for ( i = 0; i < K; i++)
        {
            numberOfVecPerGroup[i]=0;
        }
        iter++;
    }
    for (row=0; row<K; row++)
    {
        for(columns=0; columns<dim; columns++)
        {
         printf("%.4f", vecOfMeansOld[row*dim+columns]);
         if (columns!=(dim-1)){
             printf(",");
         }
        }
        printf("\n");
    }
    
    free(vector);
    free(vecOfMeansOld);
    free(numberOfVecPerGroup);
    free(vecOfMeansNew);
    return (0);
}

File: kmeans.py
#input-filepath, file is a matrix
#output-the matrix
import sys

#import from file
def fromFile():
    arr=fromFile2()
    return [split2(x.split(",")) for x in arr];

def fromFile2():
    temp=[]
    c=input()
    while (True):
        try:
            temp.append(c)
            c = input()
        except EOFError as e:
            break;
    return temp


def split2(arr):
    for i in range (len(arr)):
        arr[i]=float(arr[i]);
    return arr;



# add vector to is cluster
def add_group(vec, mean):
    for i in range(len(vec)):
        mean.append(vec[i])

# checking the distance from the mean
def distanceFromMean(n, vec, mean):
    
    d = 0
    x = 0
    distance = 0
    for i in range(n):
        meani = mean[i]
        x = vec[i]
        distance = (x - meani) * (x - meani)
        d = d + distance
    return d

# update the mean according to the last iteration
def update_mean(n, vec, mean, numOfVec):
    if numOfVec > 0:

        for i in range(n):
            meani = mean[i]
            summ = meani * numOfVec + vec[i]
            mean[i] = summ / (numOfVec + 1)
    else:
   
        for j in range(n):

            mean[j]=(vec[j])
           


def update_check(n, vec, vec2):
    for i in range(n):
        vec[i] = vec2[i]

# help function to check if we need to stop the function from runing
def to_stop(n, vec, mean):
    for i in range(n):
        if vec[i] != mean[i]:
            return False
    return True

# help function to check if we need to stop the function from runing
def to_stop_v2(d, n, a1, a2):
    for i in range(n):
        if not to_stop(d, a1[i], a2[i]):
            return False
    return True


def copyVecOfMeans(d, n, a1, a2):
    for i in range(n):
        for j in range(d):
            a2[i][j] = a1[i][j]


def mainFuncV2(k, max_iter):
    yaeri_vec= fromFile();
    iterr = 0
    d = len(yaeri_vec[0])
    number_of_vec_per_group = [0 for i in range(k)]
    vec_of_means_old = [[0 for j in range(d)] for i in range(k)]
    vec_of_means_new = [[0 for j in range(d)] for i in range(k)]
    boolean = False
    assert (k > 0)
    assert (k < len(yaeri_vec))
    assert (max_iter > 0)
    ##// initialize
    for t in range(k):
        for t1 in range(d):
            vec_of_means_old[t][t1] = yaeri_vec[t][t1]
    while iterr < max_iter:
        # distance=0
        # place=0
        cur_distance = 0
        for i in range(len(yaeri_vec)):
            distance = 0
            place = 0
            cur_distance = 0
            for j in range(k):
                vec= yaeri_vec[i]
                cur_distance = distanceFromMean(d, vec, vec_of_means_old[j])
                if j == 0:
                    distance = cur_distance
                if cur_distance < distance:
                    distance = cur_distance
                    place = j
            update_mean(d, vec, vec_of_means_new[place], number_of_vec_per_group[place])
            number_of_vec_per_group[place] = number_of_vec_per_group[place] + 1
        if to_stop_v2(d, k, vec_of_means_old, vec_of_means_new):
            boolean = True
            break
        copyVecOfMeans(d, k, vec_of_means_new, vec_of_means_old)
        number_of_vec_per_group = [0 for i in range(k)]
        iterr = iterr + 1
    for a in range(k):
        for b in range(d):
            vec_of_means_old[a][b]=format(vec_of_means_old[a][b],'.4f')
    return vec_of_means_old

def printSpecial (mat):
    for i in mat:
        for j in range(len(i)):
            print(i[j],end="")
            if (j!=len(i)-1):
                print(",",end="")
        print("\n",end="")

arr= sys.argv
if (len(arr)==2):
    x0 =int (arr[1])
    printSpecial(mainFuncV2(x0, 200))
if (len(arr)==3):
    x0=int (arr[1])
    x1= int (arr[2])
    printSpecial(mainFuncV2(x0, x1))
if (len(arr)<2):
    assert (1<0)
if (len(arr)>3):
    assert (1<0)

